Mathlib/Data/Nat/Multiplicity.lean

```diff
@@ -185,25 +186,37 @@ theorem multiplicity_choose_aux {p n b k : ℕ} (hp : p.Prime) (hkn : k ≤ n) :
     _ = _ := by simp [sum_add_distrib, sum_boole]
 #align nat.prime.multiplicity_choose_aux Nat.Prime.multiplicity_choose_aux
 
+/-- The multiplicity of `p` in `choose (n + k) k` is the number of carries when `k` and `n`
+  are added in base `p`. The set is expressed by filtering `Ico 1 b` where `b`
+  is any bound greater than `log p (n + k)`. -/
+theorem multiplicity_choose' {p n k b : ℕ} (hp : p.Prime)  (hnb : log p (n + k) < b) :
+    multiplicity p (choose (n + k) k) =
+      ((Ico 1 b).filter fun i => p ^ i ≤ k % p ^ i + n % p ^ i).card := by
+  have h₁ :
+    multiplicity p (choose (n + k) k) + multiplicity p (k ! * n !) =
+      ((Finset.Ico 1 b).filter fun i => p ^ i ≤ k % p ^ i + n % p ^ i).card +
+        multiplicity p (k ! * n !) := by
+    rw [← hp.multiplicity_mul, ← mul_assoc]
+    have := (add_tsub_cancel_right n k) ▸ choose_mul_factorial_mul_factorial (le_add_left k n)
+    rw [this, hp.multiplicity_factorial hnb, hp.multiplicity_mul,
```

Comment:
```suggestion
    rw [(add_tsub_cancel_right n k) ▸ choose_mul_factorial_mul_factorial (le_add_left k n), hp.multiplicity_factorial hnb, hp.multiplicity_mul,
```
putting the proof of `this` direcly in the rewrite here, does not seem to work, I get
"`tactic 'rewrite' failed, equality or iff proof expected`"

Reply:
`▸` can be pretty picky so this isn't that surprising.